### PR TITLE
Display status information for Deployments in Application Explorer tree #3823

### PR DESCRIPTION
**Pull request by vrubezhny**
Fixes: #3823

### CODE REVIEW

**Comment by vrubezhny** [comment]
With the latest PR update we show:
-  Deployment name,
- Number of Available Replicas,
- Actual number of created Replicas, 
- Detected `CrashLoopBackOff` state,
- Pod Container Failures (if any)
- Deployment Failures (if any)

The Deployment element icon has:
- `Green` modifier when all the Pods are successfully started and running,
- `Yellow` modifier when no `CrashLoopBackOff` state is detected, but not all the Pods are successfully started and running
- `Red` modified when either `CrashLoopBackOff` state is detected or at least one of the pods has 'Failed` or `Unknown`  state

image

File: src/explorer.ts [hunk 1]
```diff
@@ -270,24 +275,142 @@ export class OpenShiftExplorer implements TreeDataProvider<ExplorerItem>, Dispos
...
+                    messages.push(message);
+                }
+            });
+
+        return messages;
+    }
+
+    private async getDeploymentItem(element): Promise<TreeItem> {
+        const shouldHaveReplicas = element.spec.replicas > 0;
+        const availableReplicas = element.status.availableReplicas ? element.status.availableReplicas :
```
Comment (datho7561) [inline src/explorer.ts:358]: From my understanding the two numbers here would best be labelled "desiredReplicas" and "actualReplicas". "desiredReplicas" should be take from the `spec` section, and "actualReplicas" should be taken from the `status` section. Does this make sense to you? If not I might misunderstand something.

I think the difference between `status.availableReplicas` vs `status.readyReplicas` vs `status.replicas` is one means it's in the process of being created, one means it's has been created and is running, and one means it's been created and running and has a [probe](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/) configured that says it's healthy.

**Review by datho7561 — CHANGES_REQUESTED** [review]
Overall working great! Saw a few things in the code. I haven't been able to get the yellow status to show up, but it might be because it would only show up very briefly as we discussed.

(sorry it took so long to get to this).

**Comment by vrubezhny** [comment]
> I haven't been able to get the yellow status to show up, but it might be because it would only show up very briefly as we discussed.

That's simple: Just deploy your `pet-alpine` image to the Sandbox cluster - we've already discussed this - the `CrashLoopBackOff` doesn't appear on Sandbox because the Pod cannot be created, the `CreateContainerError`  is reported instead and the Pod is reported as `Pending`. So it makes it "yellow".

image

File: src/explorer.ts [hunk 2]
```diff
@@ -270,24 +275,142 @@ export class OpenShiftExplorer implements TreeDataProvider<ExplorerItem>, Dispos
...
+        pods.forEach((pod) => {
+            pod.status?.containerStatuses &&
+            pod.status.containerStatuses.forEach((cs) => {
+                if (cs.state?.waiting) {
+                    const reason = cs.state.waiting.reason;
+                    const message = cs.state.waiting.message;
+
+                    inCrashLoopBackOff = inCrashLoopBackOff || reason === 'CrashLoopBackOff';
+
+                    const msg = `${reason}: ${message ? message.trim(): 'No any valuable message'}`;
```
[the remark this answers is not in the document]
Comment (vrubezhny) [reply src/explorer.ts:326]: ACK

File: src/explorer.ts [hunk 3]
```diff
@@ -270,24 +275,142 @@ export class OpenShiftExplorer implements TreeDataProvider<ExplorerItem>, Dispos
...
+            inCrashLoopBackOff,
+            messages
+        }
+    }
+
+    private collectDeploymentErrors(deployment): string[] {
+        const messages: string[] = [];
+        deployment.status.conditions.filter((c) => c.status === 'False')
+            .forEach((c) => {
+                const message = `${c.reason}: ${c.message ? c.message.trim(): 'No any valuable message'}`;
```
[the remark this answers is not in the document]
Comment (vrubezhny) [reply src/explorer.ts:345]: ACK

File: src/explorer.ts (hunk 1 above)
Comment (vrubezhny) [reply src/explorer.ts:358]: So, OK. I take `desiredReplicas = element.spec.replicas` and don't try searching for `element.status.availableReplicas`

For the `actualReplicas` I take `element.status.replicas` value or 0 if it doesn't exist (which is possible, f.i., when scaled to 0).

But `element.status.replicas` tells nothing on real state of the replicas, as for a failing `pet-alpine` deployment (on Sandbox) we have:
```
status: {
  ...
  replicas: 1, 
  unavailableReplicas: 1, 
  updatedReplicas: 1
  ...
}
```

So, we actually have ONE replica and this the only replica is UNAVAILABLE (not working)...

... but we'll still show a tooltip like:

```
deployment: my-pet-alpine
desiredReplicas: 1
actualReplicas: 1
...
```
So user will think that everything should be OK with this replica - it's started and should be `Running`, while it's actually not (here it's `Pending`).

Probably we should show more replicas-related counters then, shouldn't we?

Comment (datho7561) [reply src/explorer.ts]: I like this behaviour!

**Review by datho7561 — APPROVED** [review]
Looks good to me. Very informative of the deployment's state. Thanks, Victor!

**Pull request merged**